File: Excel/678 Challenge.py
import math
import pandas as pd
from pandas import read_excel

path = "678 Perfect Square Even if Reversed.xlsx"
test = read_excel(path,  usecols="A:B", skiprows=1, nrows=39)

def is_square(n):
    root = int(math.isqrt(n))
    return root * root == n

def reverse_digits(n):
    return int(str(n)[::-1])

squares = [i * i for i in range(1, 40000)]
valid = [sq for sq in squares 
         if is_square(reverse_digits(sq)) and (sq % 2 != reverse_digits(sq) % 2)]

result = valid[:50]

max_len = max(len([num for num in result if num % 2 == 0]), len([num for num in result if num % 2 != 0]))
df = pd.DataFrame({
    'Even': [num for num in result if num % 2 == 0] + [None] * (max_len - len([num for num in result if num % 2 == 0])),
    'Odd': [num for num in result if num % 2 != 0] + [None] * (max_len - len([num for num in result if num % 2 != 0]))
})

print(df.equals(test))

File: Excel/600-699/678/678 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/678 Perfect Square Even if Reversed.xlsx"
test  = read_excel(path, range = "A2:B40")

is_square = function(n) {
  root = sqrt(n)
  root == floor(root)
}

reverse_digits_vec = function(n) {
  as.integer(sapply(n, function(x) paste0(rev(strsplit(as.character(x), "")[[1]]), collapse = "")))
}

n = 1:40000
squares = n^2
rev_squares = reverse_digits_vec(squares)

valid = is_square(rev_squares) & (squares %% 2 != rev_squares %% 2)
result = squares[valid][1:50] %>% 
    as_tibble() %>%
    mutate(evenodd = ifelse(value %% 2 == 0, "Even", "Odd")) %>%
    mutate(rn = row_number(), .by = evenodd) %>%
    pivot_wider(names_from = evenodd, values_from = value) %>%
    select(-rn)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE